syntax = "proto3";

/*
 This container message is send from EVerest to MCU and may contain any allowed message in that direction.
*/
message EverestToMcu {
 oneof payload {
  bool reset = 1;
  KeepAlive keep_alive = 2;
  BootConfigResponse config_response = 3;
  MeasRequest request_measurement = 4;
  bool get_calibration = 5;
  CalibrationValues set_calibration = 6;
  CalibrationRequest request_calibration = 7;
  CpState set_cp_state = 8;
  }
}

/*
 This container message is send from MCU to EVerest and may contain any allowed message in that direction.
*/
message McuToEverest {
 oneof payload {
  ResetReason reset = 1;
  KeepAlive keep_alive = 2;
  BootConfigRequest config_request = 3;
  ACMeasInstant ac_meas_instant = 4;
  ACMeasStats ac_meas_statistics = 5;
  CalibrationValues calibration_values = 6;
  CalibrationResponse response_calibration = 7;
 }
}

enum ResetReason {
	UNKNOWN = 0;
	USER = 1;
	WDG = 2;
}

message KeepAlive {
  uint32 time_stamp = 1;
  ConfigHardwareRevisionMajor hw_revision_major = 2;
  uint32 hw_revision_minor = 3;
  string sw_version_string = 4;
}

enum ConfigHardwareRevisionMajor {
  HW_MAJOR_REV_UNKNOWN = 0;
  HW_MAJOR_REV_A = 1;
}

enum MeasRequest {
  MEAS_UNKNOWN = 0;
  AC_INSTANT = 1;
  AC_STATISTICS = 2;
  DC_INSTANT = 3;
  DC_STATISTICS = 4;
  BASIC_CHARGE_COMMS_INSTANT = 5;
}

message ACMeasInstant {
  double Ua = 1;
  double Ub = 2;
  double Uc = 3;
}

message ACMeasStats {
	ACPhaseStats Ua = 1;
	ACPhaseStats Ub = 2;
	ACPhaseStats Uc = 3;
	uint32 sample_rate_ms = 4;
	uint32 window_len_samples = 5;
}

message ACPhaseStats {
	double min = 1;
	double max = 2;
	double avg = 3;
}


/* Nominal CP states */
enum CpState {
  STATE_A = 0;
  STATE_B = 1;
  STATE_C = 2;
  STATE_D = 3;
  STATE_E = 4;
  STATE_F = 5;
}

/* Nominal PP states */
enum PpState {
  STATE_NC = 0;
  STATE_13A = 1;
  STATE_20A = 2;
  STATE_32A = 3;
  STATE_70A = 4;
  STATE_FAULT = 5;
}

message BootConfigRequest {
  // TODO
}

message BootConfigResponse {
  // TODO
}

message ErrorFlags {
  // TODO
}

message CalibrationRequest {
  oneof calibration_type {
	  AC_Calibration ac = 1;
  }
}

message CalibrationResponse {
  oneof calibration_type {
	  AC_Calibration ac = 1;
  }
  uint32 value = 2;
}

message CalibrationValues {
  uint32 atm90_ua_gain = 1;
  uint32 atm90_ub_gain = 2;
  uint32 atm90_uc_gain = 3;
  uint32 atm90_ua_offset = 4;
  uint32 atm90_ub_offset = 5;
  uint32 atm90_uc_offset = 6;
}

enum ACPhase {
	Ua = 0;
	Ub = 1;
	Uc = 2;
}

message AC_Calibration {
	ACPhase phase = 1;
	oneof type {
		bool offset = 2;  // unused for now, MCU will detect calibration on offset by which_type
		double gain = 3;  // use this value as reference voltage passed to calibration routine
	}
}